libcache/test_libcache.c

```diff
@@ -0,0 +1,989 @@
+	}
+ */
+/* cases: 	true, true -> true (short-circuit)
+			true, false -> true (short-circuit)
+			false, true -> true (no short-circuit)
+			false, false -> false (no short-circuit)
+*/
+
+/* true, true -> true (short-circuit) */
+TEST(test_init, cache_init_size_mod_line_size_not_zero)
+{
+	size_t size = 1754, lineSize = 17; /* size % lineSize = 3, (int)(size / lineSize) % LIBCACHE_NUM_WAYS = 3 */
+	cachectx_t *cache = NULL;
+
+	cache = cache_init(LIBCACHE_SRC_MEM_SIZE, size, lineSize, test_writeCb, test_readCb);
+	TEST_ASSERT_NULL(cache);
+}
+
+/* false, true -> true (no short-circuit) */
+TEST(test_init, cache_init_size_mod_line_size_zero)
+{
+	size_t size = 1026, lineSize = 19; /* size % lineSize = 0, (size / lineSize) % LIBCACHE_NUM_WAYS = 2 */
+	cachectx_t *cache = NULL;
+
+	cache = cache_init(LIBCACHE_SRC_MEM_SIZE, size, lineSize, test_writeCb, test_readCb);
+	TEST_ASSERT_NULL(cache);
+}
+
+/* false, false -> false (no short-circuit) */
+/* covers both ifs */
+TEST(test_init, cache_init_sizes_not_zero_bytes)
+{
+	int ret = -1;
+	size_t size = LIBCACHE_CACHE_SIZE, lineSize = LIBCACHE_LINE_SIZE;
+	cachectx_t *cache = NULL;
+
+	cache = cache_init(LIBCACHE_SRC_MEM_SIZE, size, lineSize, test_writeCb, test_readCb);
+	TEST_ASSERT_NOT_NULL(cache);
+
+	ret = cache_deinit(cache);
+	TEST_ASSERT_EQUAL_INT(EOK, ret);
+}
+
+TEST_GROUP_RUNNER(test_init)
+{
+	RUN_TEST_CASE(test_init, cache_init_sizes_zero_bytes);
+	RUN_TEST_CASE(test_init, cache_init_line_size_zero_bytes);
+	RUN_TEST_CASE(test_init, cache_init_size_mod_line_size_not_zero);
+	RUN_TEST_CASE(test_init, cache_init_size_mod_line_size_zero);
+	RUN_TEST_CASE(test_init, cache_init_sizes_not_zero_bytes);
+}
+
+TEST_GROUP(test_deinit);
+
+TEST_SETUP(test_deinit)
+{
+}
+
+TEST_TEAR_DOWN(test_deinit)
+{
+}
+
+TEST(test_deinit, cache_deinit_uninitialized_cache)
+{
+	int ret;
+	cachectx_t *cache = NULL;
+
+	ret = cache_deinit(cache);
+	TEST_ASSERT_EQUAL_INT(-EINVAL, ret);
+}
+
+TEST(test_deinit, cache_deinit_initalized_cache)
+{
+	int ret;
+	cachectx_t *cache;
+
+	cache = cache_init(LIBCACHE_SRC_MEM_SIZE, LIBCACHE_CACHE_SIZE, LIBCACHE_LINE_SIZE, test_writeCb, test_readCb);
+	TEST_ASSERT_NOT_NULL(cache);
+
+	ret = cache_deinit(cache);
+	TEST_ASSERT_EQUAL_INT(EOK, ret);
+}
+
+TEST_GROUP_RUNNER(test_deinit)
+{
+	RUN_TEST_CASE(test_deinit, cache_deinit_uninitialized_cache);
+	RUN_TEST_CASE(test_deinit, cache_deinit_initalized_cache);
+}
+
+TEST_GROUP(test_read_write);
+
+TEST_SETUP(test_read_write)
+{
+	int ret;
+
+	fileRead = open("/var/read.txt", O_RDONLY | O_CREAT);
+	TEST_ASSERT_NOT_EQUAL_INT(-1, fileRead);
+
+	fileWrite = open("/var/write.txt", O_RDWR | O_CREAT);
+	TEST_ASSERT_NOT_EQUAL_INT(-1, fileWrite);
+
+	offBitsNum = LOG2(LIBCACHE_LINE_SIZE);
+	offMask = ((uint64_t)1 << offBitsNum) - 1;
+}
+
+TEST_TEAR_DOWN(test_read_write)
+{
+	int ret = -1;
+
+	ret = close(fileRead);
+	TEST_ASSERT_EQUAL_INT(0, ret);
+
+	ret = close(fileWrite);
+	TEST_ASSERT_EQUAL(0, ret);
+
+	ret = remove("/var/write.txt");
```

Comment:
In the tests, `/var/write.txt` and `/var/read.txt` simulate a source memory device that the `libcache` library can write to and read from by the use of suitable callbacks. Here the write and read operations are 'separated' and they are performed on two different files unlike in real settings, where they happen on the same source memory. This approach has helped me to empirically evaluate the outcome of the operations, especially the write operation.

I think I may need to 'merge' the files to simulate the source memory more accurately. This would also reduce the overhead caused by reoccurring creation of additional files.

Reply:
This above is no longer an issue as I've merged the files into one.